rust/arrow/src/array/data.rs

```diff
@@ -467,6 +467,37 @@ impl ArrayData {
     }
 }
 
+impl ArrayData {
+    /// Creates a new [`ArrayData`] for a primitive array, verifying all assumptions for consumers to
+    /// soundly use it.
+    /// # Panic
+    /// This function panics iff any of the following:
+    /// * `values.len()` is not a multiple of `size_of::<T::Native>`
+    /// * `values.as_ptr()` is not aligned with `T::Native`
+    /// * when `nulls` is `Some`, `nulls.len` is smaller than `(ArrayData::len + 7) / 8`.
+    #[inline]
+    pub fn new_primitive<T: ArrowPrimitiveType>(
+        values: Buffer,
+        nulls: Option<Buffer>,
```

Comment:
Should this cater for offsets, or is your preference to omit it? If we took an offset, we'd add an extra assertion that checks that `offset < calculated_len`.

Thinking more about it, if we don't take an offset, we might force users to still use `ArrayData::new` for that facility, which might be undesirable.